keras/callbacks.py

```diff
@@ -834,41 +836,20 @@ def is_indexed_slices(grad):
 
             projector.visualize_embeddings(self.writer, config)
 
+    def on_epoch_begin(self, epoch, logs=None):
+        if self.model.test_function is None:
+            return
+        self.model.test_function._epoch = epoch
+        self.model.test_function._writer = self.writer
```

Comment:
It adds a lot of flat complexity to add all these attributes to Functions. Better to have a TensorBoard API on Function, like `_set_tensorboard_config(merged_summaries, writer, epoch)`. In general I'm not a fan of splitting TB functionality across two unrelated objects (TB callback and Function). Who's calling `writer.flush()`?